unity/Assets/Scripts/AgentManager.cs

```diff
@@ -728,147 +704,184 @@ private void addImageSynthesisImage(List<KeyValuePair<string, byte[]>> payload,
-                        break;
+                    // read body
+                    while (bodyBytesReceived < contentLength) {
+                        // check for 0 bytes received
+                        int received = this.sock.Receive(bodyBuffer, bodyBytesReceived, bodyBuffer.Length - bodyBytesReceived, SocketFlags.None);
+                        if (received == 0) {
+                            Debug.LogError("0 bytes received attempting to read body - connection closed");
+                            break;
+                        }
+
+                        bodyBytesReceived += received;
+                        //Debug.Log("total bytes received: " + bodyBytesReceived);
                     }
 
-                    bodyBytesReceived += received;
-                    //Debug.Log("total bytes received: " + bodyBytesReceived);
+                    string msg = Encoding.ASCII.GetString(bodyBuffer, 0, bodyBytesReceived);
+                    ProcessControlCommand(msg);
                 }
-
-                string msg = Encoding.ASCII.GetString(bodyBuffer, 0, bodyBytesReceived);
+            } else if (serverType == serverTypes.FIFO){
+                byte[] msgPackMetadata = MessagePack.MessagePackSerializer.Serialize(multiMeta, 
+                    MessagePack.Resolvers.ThorContractlessStandardResolver.Options);
+                this.fifoClient.SendMessage(FifoServer.FieldType.Metadata, msgPackMetadata);
+                foreach(var item in renderPayload) {
+                    this.fifoClient.SendMessage(FifoServer.Client.FormMap[item.Key], item.Value);
+                }
+                this.fifoClient.SendEOM();
+                string msg = this.fifoClient.ReceiveMessage();
                 ProcessControlCommand(msg);
-            }
-		} else if (serverType == serverTypes.FIFO){
-            byte[] msgPackMetadata = MessagePack.MessagePackSerializer.Serialize(multiMeta, 
-            MessagePack.Resolvers.ThorContractlessStandardResolver.Options);
-            this.fifoClient.SendMessage(FifoServer.FieldType.Metadata, msgPackMetadata);
-            foreach(var item in renderPayload) {
-                this.fifoClient.SendMessage(FifoServer.Client.FormMap[item.Key], item.Value);
-            }
-            this.fifoClient.SendEOM();
-            string msg = this.fifoClient.ReceiveMessage();
-            ProcessControlCommand(msg);
-		}
 
-        if(droneMode)
-        {
-            if (Time.timeScale == 0 && !Physics.autoSimulation && physicsSceneManager.physicsSimulationPaused)
-            {
-                DroneFPSAgentController agent_tmp = this.agents.ToArray()[0].GetComponent<DroneFPSAgentController>();
-                Time.timeScale = agent_tmp.autoResetTimeScale;
-                Physics.autoSimulation = true;
-                physicsSceneManager.physicsSimulationPaused = false;
-                agent_tmp.hasFixedUpdateHappened = false;
+                while (canEmit() && this.fastActionEmit) {
```

Comment:
What happens for the case where `fastActionEmit` is passed as true to the action but the controller is not using the `fifoServer` ?

Reply:
Everything works as expected, you just don't get any speed improvements.  I just added a warning in controller.py so that users are aware that it only works with the FifoServer.